core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/PartialSuccessHandler.java

```diff
@@ -0,0 +1,192 @@
+	 * @return Returns a pass/fail for handling this transmission.
+	 */
+	boolean validateTransmissionAndSend(TransmissionHandlerArgs args) {
+		if (args.getTransmission() != null && args.getTransmissionDispatcher() != null) {
+			switch (args.getResponseCode()) {
+			case HttpStatus.SC_PARTIAL_CONTENT:
+				BackendResponse backendResponse = getBackendResponse(args.getResponseBody());
+				List<String> originalItems = generateOriginalItems(args);
+
+				// Somehow the amount of items received and the items sent do not match
+				if (backendResponse != null && (originalItems.size() != backendResponse.itemsReceived)) {
+					InternalLogger.INSTANCE.trace(
+							"Skipping partial content handler due to itemsReceived being larger than the items sent.");
+					return false;
+				}
+
+				if (backendResponse != null && (backendResponse.itemsAccepted < backendResponse.itemsReceived)) {
+					List<String> newTransmission = new ArrayList<String>();
+					for (BackendResponse.Error e : backendResponse.errors) {
+						switch (e.statusCode) {
+						case TransmissionSendResult.REQUEST_TIMEOUT:
+						case TransmissionSendResult.INTERNAL_SERVER_ERROR:
+						case TransmissionSendResult.SERVICE_UNAVAILABLE:
+						case TransmissionSendResult.THROTTLED: 
+						case TransmissionSendResult.THROTTLED_OVER_EXTENDED_TIME: 
+							// Unknown condition where backend response returns an index greater than the
+							// items we're returning
+							if (e.index < originalItems.size()) {
+								newTransmission.add(originalItems.get(e.index));
+							}
+							break;
+						}
+					}
+					return sendNewTransmission(args, newTransmission);
+				}
+				InternalLogger.INSTANCE
+						.trace("Skipping partial content handler due to itemsAccepted and itemsReceived being equal.");
+				return false;
+
+			default:
+				InternalLogger.INSTANCE.trace("Http response code %s not handled by %s", args.getResponseCode(),
+						this.getClass().getName());
+				return false;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Used to parse the original telemetry request in order to resend the failed
+	 * ones.
+	 * 
+	 * @param args
+	 *            The {@link TransmissionHandlerArgs} that contains the
+	 *            {@link Transmission} object.
+	 * @return A List<> of each sent item
+	 */
+	List<String> generateOriginalItems(TransmissionHandlerArgs args) {
+		List<String> originalItems = new ArrayList<String>();
+
+		if (args.getTransmission().getWebContentEncodingType() == "gzip") {
+
+			try {
+				GZIPInputStream gis = new GZIPInputStream(
+						new ByteArrayInputStream(args.getTransmission().getContent()));
+				BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(gis));
+				String line;
+				while ((line = bufferedReader.readLine()) != null) {
+					originalItems.add(line);
+				}
+				if (gis != null) {
+					gis.close();
+				}
+				if (bufferedReader != null) {
+					bufferedReader.close();
+				}
```

Comment:
This should be closed in finally statement?

Reply:
Yes this if should definitely go in finally!

Reply:
lines 125--130 should be wrapped in the finally block below

Reply:
Yup, moved to finally and added try/catch around those close() statements as well.